src/main/java/io/reactivex/internal/operators/flowable/FlowableBufferBoundary.java

```diff
@@ -98,214 +119,325 @@ public void onSubscribe(Subscription s) {
+            } catch (Throwable ex) {
+                Exceptions.throwIfFatal(ex);
+                SubscriptionHelper.cancel(upstream);
+                if (errors.addThrowable(ex)) {
+                    subscribers.dispose();
+                    synchronized (this) {
+                        buffers = null;
+                    }
+                    done = true;
+                    drain();
+                } else {
+                    RxJavaPlugins.onError(ex);
+                }
                 return;
             }
 
-            U b;
-
-            try {
-                b = ObjectHelper.requireNonNull(bufferSupplier.call(), "The buffer supplied is null");
-            } catch (Throwable e) {
-                Exceptions.throwIfFatal(e);
-                onError(e);
-                return;
+            long idx = index;
+            index = idx + 1;
+            synchronized (this) {
+                Map<Long, C> bufs = buffers;
+                if (bufs == null) {
+                    return;
+                }
+                bufs.put(idx, buf);
             }
 
-            Publisher<? extends Close> p;
+            BufferCloseSubscriber<T, C> bc = new BufferCloseSubscriber<T, C>(this, idx);
+            subscribers.add(bc);
+            p.subscribe(bc);
+        }
 
-            try {
-                p = ObjectHelper.requireNonNull(bufferClose.apply(window), "The buffer closing publisher is null");
-            } catch (Throwable e) {
-                Exceptions.throwIfFatal(e);
-                onError(e);
-                return;
+        void openComplete(BufferOpenSubscriber<Open> os) {
+            subscribers.delete(os);
+            if (subscribers.size() == 0) {
+                SubscriptionHelper.cancel(upstream);
+                done = true;
+                drain();
             }
+        }
 
-            if (cancelled) {
-                return;
+        void close(BufferCloseSubscriber<T, C> closer, long idx) {
+            subscribers.delete(closer);
+            boolean makeDone = false;
+            if (subscribers.size() == 0) {
+                makeDone = true;
+                SubscriptionHelper.cancel(upstream);
             }
-
             synchronized (this) {
-                if (cancelled) {
+                Map<Long, C> bufs = buffers;
+                if (bufs == null) {
                     return;
                 }
-                buffers.add(b);
+                queue.offer(buffers.remove(idx));
             }
+            if (makeDone) {
+                done = true;
+            }
+            drain();
+        }
 
-            BufferCloseSubscriber<T, U, Open, Close> bcs = new BufferCloseSubscriber<T, U, Open, Close>(b, this);
-            resources.add(bcs);
+        void boundaryError(Disposable subscriber, Throwable ex) {
+            SubscriptionHelper.cancel(upstream);
+            subscribers.delete(subscriber);
```

Comment:
Can we just `dispose()` all subscribers here and remove `dispose()` call below?

Reply:
Same in Observable version

Reply:
It avoids disposing the caller which is known to have reached its terminal state.